### PR TITLE
Feature/restructure encryption

### PR DESCRIPTION
**Pull request by lfarrel6**
- tidied encrypt decrypt functions
- added options parameter

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,269 +1,291 @@
...
-            return handleEncryption(data, Object.keys(data));
-          });
-        }
-    
-        return handleEncryption(data, Object.keys(data));
+          return encryptedObject;
+        };
+
+        return handleEncryption(data, fields);
       } else {
```
Comment (eoinboylan) [inline index.js]: Else is redundant just return

File: index.js [hunk 2]
```diff
@@ -1,269 +1,291 @@
...
-            dataBuffer
-          ).then(decryptedData => {
-            return utils.bufferToString(decryptedData);
-          }).catch(error => {
-            throw new Error(error)
-          });
+
+    _encryptArray: function(data) {
+      const handleEncryption = async (array) => {
+        let encryptedArr = new Array(array.length);
```
Comment (eoinboylan) [inline index.js]: I'd make this use a map instead of for if its just returning a new array. Not blocker.

File: index.js [hunk 3]
```diff
@@ -1,269 +1,291 @@
...
-            return handleDecryption(data);
+        return encryptedArr;
+      };
+
+      return handleEncryption(data);
+    },
+
+    /**
+     * @param data: string or object containing the data to be encrypted
+     * @param options: see evervault developer docs
```
Comment (eoinboylan) [inline index.js]: Side note, we should create a typings file for the SDK so devs don't have to leave the IDE to see options options types.

File: index.js [hunk 4]
```diff
@@ -1,269 +1,291 @@
...
-    
-        const handleDecryption = async (object,fieldsToDecrypt) => {
-          let decryptedObject = Object.assign({},object);
-          for (let index = 0; index < fieldsToDecrypt.length; index++) {
-            if (object[fieldsToDecrypt[index]]) {
-              const parsedEncData = utils.parseEvervaultString(object[fieldsToDecrypt[index]]);
-              const dataBuffer = utils.b64ToBuffer(parsedEncData.browserData);
-              const ivBuffer = utils.b64ToBuffer(parsedEncData.iv);
-              decryptedObject[fieldsToDecrypt[index]] = await decryptField(dataBuffer, ivBuffer);
+        let fields = options.fieldsToEncrypt || Object.keys(data);
```
Comment (eoinboylan) [inline index.js]: const

File: index.js [hunk 5]
```diff
@@ -1,269 +1,291 @@
...
+                JSON.stringify(decryptedObject[key])
+              );
             }
           }
-          return decryptedObject
+          return decryptedObject;
+        };
+
+        return handleDecryption(data, fields);
+      } else {
```
Comment (eoinboylan) [inline index.js]: redundant else

File: index.js [hunk 6]
```diff
@@ -1,269 +1,291 @@
...
+        if (data.constructor.name === 'Array') {
+          return this.getKey(options.privateKey).then(() => {
+            return this._decryptArray(data);
+          });
         }
-    
-        return handleDecryption(data, Object.keys(data));
-      } else {
-        throw new Error("evervault error: data must be a string or an object");
+        let fields = options.fieldsToDecrypt || Object.keys(data);
```
Comment (eoinboylan) [inline index.js]: const

File: index.js [hunk 7]
```diff
@@ -1,269 +1,291 @@
...
       return new Promise((resolve, reject) => {
         const urlKey = window.location.hash.substring(2);
-        if(urlKey){
-  
+        if (urlKey) {
           const params = urlKey.split(':');
           resolve({
             privateKey: params[0],
-            authToken: params[1]
+            authToken: params[1],
```
Comment (eoinboylan) [inline index.js]: will now need refresh token to

File: index.js [hunk 8]
```diff
@@ -1,269 +1,291 @@
...
   },
 
-  buildEvervaultString: function(iv,encoded){
+  buildEvervaultString: function(iv, encoded) {
     return `enc:v1:${utils.bufferToB64(iv)}:${utils.bufferToB64(encoded)}`;
   },
-  
+
   setUserKeysInStorage: function(urlKey) {
     const keys = urlKey.split(':');
```
Comment (eoinboylan) [inline index.js]: set refresh token

File: index.js [hunk 9]
```diff
@@ -1,269 +1,291 @@
...
-    logout: function () {
-      localStorage.removeItem('evervault-privateKey')
-      localStorage.removeItem('evervault-authToken')
-      localStorage.removeItem('evervault-haiku')
-      window.location.replace(`${authRedirectUrl}/${MACHINE_NAME}`)
+    logout: function() {
+      localStorage.removeItem('evervault-privateKey');
+      localStorage.removeItem('evervault-accessToken');
+      localStorage.removeItem('evervault-haiku');
+      window.location.replace(`${authRedirectUrl}/${MACHINE_NAME}`);
```
Comment (eoinboylan) [inline index.js]: remove refresh token

**Comment by lfarrel6** [comment]
needs further testing, will reopen

**Pull request closed without merging**